File: kqbc.py
import numpy as np
from scipy.linalg import schur
from sklearn.metrics.pairwise import rbf_kernel
import matlab.engine
import copy



def hit_n_run(x, A ,T):
    """
    Returns a random point using the hit and run algorithm from the
    convex body defined by Ax>=0 and ||x||<=1.
    The random walks begins from the point x which is assumed
    to be an internal point (i.e. satisfies the constraints Ax>=0
    and ||x||<=1. The number of steps the algorithms will perform
    is T.

    Inputs:
    x - A starting point for the random walk. Must be internal
       point.

    A - A set of constraints defining the convex body: Ax>=0

    T - Number of steps to perform.

    Outputs:
    x - A random point in the convex body {x: Ax>=0, ||x||<=1}.

    """


    dim = len(x)
    x = x[:]
    u = np.random.randn(T, dim) # at step t the algorithm will pick a random point
                               # on the line through x and x+u(t,:)
    # u = np.arange(1, T * dim + 1).reshape(T, dim, order='F')
    Au = u @ A.T
    nu = np.sum(u**2, 1)
    l = np.random.rand(T, 1)
    # l = np.arange(1, T + 1)

    for t in np.arange(0, T):
        Ax = A @ x
        ratio = -np.divide(Ax, Au[t, :][np.newaxis].T)
        I = np.where(Au[t,:] > 0)[0]
        mn = np.append(ratio[I], -np.inf).max()
        I = np.where(Au[t, :] < 0)[0]
        mx = np.append(ratio[I], np.inf).min()

        disc = ((x.T @ u[t, :].T) ** 2 - nu[t] * (np.linalg.norm(x) ** 2 - 1))[0]
        if disc < 0:
            print('negative disc, probably x is not a feasable point.')
            disc = 0

        hl = (-(x.T @ u[t, :][np.newaxis].T) + np.sqrt(disc)) / nu[t]
        ll = (-(x.T @ u[t, :][np.newaxis].T) - np.sqrt(disc)) / nu[t]

        xx = min(hl[0], mx)
        nn = max(ll[0], mn)
        x = x + u[t, :][np.newaxis].T * (nn + l[t] * (xx-nn))
    return x




def KQBC(K, Y_train, T):
    """
    Runs the kernelized version of the Query By Committee (QBC) algorithm.

    Inputs:
        K       - Training instances with kernel applied.
        Y_train - Training labels (values +1 and -1).
        T       - Number of random walk steps to make when selecting a
                 random hypothesis.
        kernel  - Type of kernel to be used. Possible values are
                 'Linear', 'Poly' and 'Gauss'.
        parameter1, parameter2 -
                 parameters for the chosen kernel.

    Outputs:
        selection - The instances for which the algorithm queried for
                   label.
        coefs     - The coefficients of the hypotheses used in each of
                   the steps of the algorithm.
        errors    - The training error at each step.
    """

    tol = 1e-10 #tolerance for the factor function

    coefs = []
    errors = []

    samp_num = len(Y_train)

    selection = [0] # initialization: select the first sample point
    selected = 1

    coef = np.zeros((len(Y_train), 1))
    coef[0] = Y_train[0] / np.sqrt(K[0, 0])
    # preds = K @ coef
    # errate = sum(np.multiply(Y_train.squeeze(), preds.squeeze()) <= 0)

    for ii in np.arange(1, samp_num):
        extension = selection + [ii]
        s, u = schur(K[extension][: , extension])
        s = np.diag(s)
        I = np.where(s > tol)[0]
        A = u[:, I] @ np.diag(s[I] ** -0.5)

        y_select_diag = Y_train[selection] if len(selection) == 1 else np.diag(Y_train[selection, 0])
        restri = y_select_diag @ K[selection][: , extension] @ A

        co1 = np.linalg.pinv(A) @ coef[extension]

        # try:
        co2 = hit_n_run(co1, restri, T)
        co1 = hit_n_run(co1, restri, T)

        pred1 = K[ii, extension] @ (A @ co1)
        pred2 = K[ii, extension] @ (A @ co2)

        if (pred1 * pred2 <= 0): # the classifiers disagree
            selection = extension
            if (Y_train[ii] * pred1 >= 0):
                coef[extension] = A @ co1
            else:
                coef[extension] = A @ co2

            coefs.append(copy.deepcopy(coef))
            errors.append(np.sum(np.multiply(Y_train, K @ coef) <= 0) / len(Y_train))
            # print((len(selection), errors[-1]))

    return selection, coefs, errors

File: matlab/KQBC.m
function [selection, ret_w] = KQBC(X_train, Y_train, T, kernel, queries, ...
                                           parameter1,parameter2)
% function [selection, coefs, errors] = KQBC(X_train, Y_train, queries, ...
%                                            T,kernel, parameter1,parameter2)
%
% Runs the kernelized version of the Query By Committee (QBC) algorithm.
%
% Inputs:
%   X_train - Training instances.
%   Y_train - Training labels (values +1 and -1).
%   T       - Number of random walk steps to make when selecting a
%             random hypothesis.
%   kernel  - Type of kernel to be used. Possible values are
%             'Linear', 'Poly' and 'Gauss'.
%   queries  - Number of queries indicate for the loop to break.
%   parameter1, parameter2 -
%             parameters for the chosen kernel.
%
% Outputs:
%   selection - The instances for which the algorithm queried for
%               label.
%   coefs     - The coefficients of the hypotheses used in each of
%               the steps of the algorithm.
%   errors    - The training error at each step.


% Copyright (C) 2005 Ran Gilad-Bachrach
%
% This program is distributed in the hope that it will be useful,
% but WITHOUT ANY WARRANTY; without even the implied warranty of
% MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
%
% Change Log
% ----------
% Version 1.0 - May 4 2005



tol = 1e-10;   %tolerance for the factor function


  switch kernel
    case 'Linear'
      K = X_train*X_train';
    case 'Poly'
      K = (X_train*X_train' + parameter1).^parameter2;
    case 'Gauss'
      nrm = sum(X_train.^2,2);
      K = exp(-(repmat(nrm,1,length(Y_train))-2*X_train*X_train'+...
		repmat(nrm',length(Y_train),1))/(2*parameter1.^2));
    otherwise
      disp(['unknown kernel']);
      return;
  end

  coefs=[];
  errors=[];
  ret_w=[];

  samp_num = length(Y_train);


  selection = [1];    % initialization: select the first sample point
  selected = 1;

  coef = zeros(length(Y_train),1);
  coef(1) = Y_train(1)/sqrt(K(1,1));
  preds = K*coef;
  errate = sum(Y_train.*preds<=0);

  for ii= 2:samp_num;
    if length(selection) >= queries
        break;
    end
    extension = [selection ii];
    % the following steps correspond to finding an orthonormal
    % basis to the sample points. We generate a matrix A such that
    % A'*K(extension,extension)*A is the identity matrix with rank
    % equals to the rank of K.
    [u,s] = schur(K(extension,extension));
    s = diag(s);
    I = (s>tol);
    A = u(:,I)*diag(s(I).^-0.5);

    restri = diag(Y_train(selection))*K(selection,extension)*A;

    co1 = pinv(A)*coef(extension);


    co2 = hit_n_run(co1,restri,T);
    co1 = hit_n_run(co2,restri,T);

    pred1 = K(ii,extension)*(A*co1);
    pred2 = K(ii,extension)*(A*co2);

    if (pred1*pred2<=0)   % the classifiers disagree
      selection = extension;
      if (Y_train(ii)*pred1>=0)
	    coef(extension) = A*co1;
        ret_w = co1;
      else
	    coef(extension) = A*co2;
        ret_w = co2;
      end
      coefs = [coefs,coef];
      errors = [errors;sum(Y_train.*(K*coef)<=0)/length(Y_train)];
      %disp([ii,length(selection),errors(end)]);
    end
  end
%end

